### PR TITLE
Add Opera for Android

### PR DESCRIPTION
**Pull request by ExE-Boss**
Part of #591

**Note:** This might not actually be fully accurate, as I don’t use Opera.

review?(@Elchi3, @jpmedley)

### CODE REVIEW

**Comment by ExE-Boss** [comment]
review?(@Elchi3)

**Comment by Elchi3** [comment]
Oh and I think in this case it doesn't make sense to link to Opera Desktop release notes as it is not the same.

**Comment by jpmedley** [comment]
It looks like I missed responding to a request for review. I do not have access to any information beyond the sources the Forlian has already consulted.

**Comment by ExE-Boss** [comment]
I plan to look into it during the weekend.

**Comment by vinyldarkscratch** [comment]
Moving the discussion to this PR.  What was discussed in #2909 is as follows:

[Comment 1](https://github.com/mdn/browser-compat-data/pull/2909#issuecomment-426862682)
> Looks like we have a little bit of an inconsistency on our hands in terms of the version numbers. Opera Android has different version numbers than Opera Desktop, that’s already been established ~~in the discussion on #1712~~ up above.
> 
> https://www.apk4fun.com/history/998/
> https://en.wikipedia.org/wiki/Opera_Mobile
> 
> However, ChromeStatus.com seems to be referring to Opera Android version numbers as if they were the same as the desktop version numbers, as well as referring to versions way beyond the current released version.
> 
> https://www.chromestatus.com/feature/5718547946799104
> 
> So...it must be referring to the underlying engine in Opera Android, which follows the desktop version’s numbers?

[Comment 2](https://github.com/mdn/browser-compat-data/pull/2909#issuecomment-427059061)
> I think the version numbers that I [@Elchi3] mention here #1712 (comment) are correct. Whether or not that's actually doable and makes sense for our data, I don't know yet. Strictly speaking chromestatus seems to do the wrong thing, but maybe they refer to some engine versions. I haven't had the time to dig deeper into this yet.

[Comment 3](https://github.com/mdn/browser-compat-data/pull/2909#issuecomment-427066211)
> As far as version numbers, a similar case is evident with Microsoft Edge, as described in #2620, where the engine and the application versions are different. As also described, the engine version is preferred, rather than the application version. Personally, I feel that maybe we should keep with that regarding Opera Android (especially since I expect people would believe chromestatus.com's accuracy, and it would also be a little less work :P). But it's all up to you guys in the end, what you think would be best!

**Comment by jpmedley** [comment]
"Looks like we have a little bit of an inconsistency on our hands in terms of the version numbers. Opera Android has different version numbers than Opera Desktop, that’s already been established in the discussion on #1712 up above."

According to one of Chrome's Developer Advocates who worked for Opera until a few years ago:

The wikipedia article is incorrect. Opera for Android followed the OPERA_VERSION = CHROMIUM_VERSION - 13 formula, just like Opera for Desktop. However, it looks like this recently changed. The current Opera for Android version is v47~, but its Chromium version is v66~.

Opera did not tell us.

**Comment by Elchi3** [comment]
My wild guess is that "OPERA_VERSION = CHROMIUM_VERSION - 13" worked until Opera Android 37. I think it would be great to find out if there is a new formula for the later versions.

41, 2016-12-22, based on Chromium v?
42, 2017-02-20, based on Chromium v?
43, 2017-09-27, based on Chromium v?
44, 2017-12-11, based on Chromium v?
45, 2018-02-14, based on Chromium v?
46, 2018-05-14, based on Chromium v?
47, 2018-07-23, based on Chromium v66
48, 2018-??-??, based on Chromium v?

Anyone knows how to find out?

**Comment by mathiasbynens** [comment]
Opera employees should know.

**Comment by Elchi3** [comment]
@chrisdavidmills Do you know Opera employees that could help us here?

**Comment by chrisdavidmills** [comment]
@Elchi3 I don't know anyone from Opera any more; everyone I know left.

@mathiasbynens was there much more recently than me; so was Andres Bovens.

**Comment by chrisdavidmills** [comment]
@Elchi3 I don't know anyone from Opera any more; everyone I know left.

@mathiasbynens was there much more recently than me; so was Andres Bovens.

**Comment by poshaughnessy** [comment]
In case a few more are helpful in the meantime...

* v45 = Chromium v61 (found from Play Store description [from Internet Archive](https://web.archive.org/web/20180505040517/https://play.google.com/store/apps/details?id=com.opera.browser))
* v46 = Chromium v63 (found from [this Opera Forums announcement](https://forums.opera.com/topic/26662/opera-for-android-46), NB. possibly you could try contacting [Miyuki there via the forum](https://forums.opera.com/user/miyukiwork)?)
* v47 = Chromium v66 (found from `opera://version/` using current Stable)
* v48 = Chromium v69 (found from `opera://version/` using current Beta)

**Comment by Elchi3** [comment]
Thanks @poshaughnessy! :+1: Especially the forum link was super helpful! I've my comment above (https://github.com/mdn/browser-compat-data/pull/1712#issuecomment-416886047) with what we know now.

**Comment by ExE-Boss** [comment]
review?(@Elchi3): I would like to get this merged ASAP and fix it up later once we get new information for https://github.com/mdn/browser-compat-data/pull/1712#issuecomment-416886047.

**Comment by Elchi3** [comment]
Just to make expectations clear: Since this is the last browser without _any_ validation, I'd like to get to this, too, but I can't review it ASAP (whatever that means anyways) as it is a large change and not a browser that is currently a priority.

Also, I thought that maybe https://github.com/mdn/browser-compat-data/issues/3264 and a test that uses that info would help with getting this done correctly.

**Comment by odinho** [comment]
TL;DR - nothing of relevance, talking about Opera desktop versions

------

I can tell you something that might help understand *some* things about Opera version numbers (which was made to use same system as Chromium, originally, before different people changed things up in different ways for unknown reasons).

So if you know Chromium's versioning, you'd know Opera's original versioning too.

The version number is `<marketing_version>.<variant_version>.<nightly_version>.<official_branch_build>`.

So `marketing_version` is the version we're talking about here. For desktop this is `chromium - 13`, for Android they've fscked it up for a random reason. Probably they don't want to show that they've been slow in upgrading to newer Chromium (typical marketing reason).

`variant_version` would for desktop always be 0, before I quit we had just started to use it for AB-test versions of Opera though. So you could see if you had a special build if that version was not 0.

`nightly_version` is the *real* version we actually cared about internally. It's a number that increases once every night. A nightly build then gets made, which gets a number like `20.0.2239.0`, the patch 0 always means this is the nightly build. This is often what we'd release in the "developer" channel. When we decided to branch out for a new stable it would get its own branch and every time we did an official build on that branch, the last `0` would increase. Branch versions could only go `1-999` I think at the start, `1000+` was for manual builds done on that nightly version, and `10000+` was for any random build, and continous integration builds got a new number stuck on at the end that was monotically increasing from `0` to `99999` with no nightly reset (the others were reset nightly).

All this is little help. Except the `nightly_version` can tell you at what date that particular version was branched off from the internal Opera `master` branch. So if nothing more, it lets you calculate how many days was spent between released versions :) :man_shrugging:

**Comment by ExE-Boss** [comment]
I’d like to get this in before another merge conflict arises.

**Comment by Elchi3** [comment]
How can I review this? Many Opera for Android values that were wrongly based off of the formula Chromium - 13 are affected here, right? How did you make changes? Did you use a script that takes the mapping from https://github.com/mdn/browser-compat-data/pull/1712#issuecomment-416886047 into account?

**Comment by ExE-Boss** [comment]
> How can I review this? Many Opera for Android values that were wrongly based off of the formula Chromium - 13 are affected here, right? How did you make changes? Did you use a script that takes the mapping from https://github.com/mdn/browser-compat-data/pull/1712#issuecomment-416886047 into account?

Yes, that’s exactly what I did.

**Comment by Elchi3** [comment]
How does the script work? Does it update all opera_android versions based on chrome_android and the mapping?

File: api/AudioBufferSourceNode.json [hunk 1]
```diff
@@ -393,7 +393,7 @@
             },
             "opera_android": {
               "version_added": true,
-              "version_removed": "44"
+              "version_removed": "10.1"
```
Comment (Elchi3) [inline api/AudioBufferSourceNode.json]: I don't understand this change from 44 to 10.1. Can you explain?

File: api/AudioContext.json [hunk 2]
```diff
@@ -55,7 +55,7 @@
             },
             {
               "version_added": "15",
-              "version_removed": "44",
+              "version_removed": "15",
```
Comment (Elchi3) [inline api/AudioContext.json]: 43 should be correct according to the mapping, I believe. Added and removed both in 15 makes no sense.

File: api/AudioListener.json [hunk 3]
```diff
@@ -85,7 +85,7 @@
             },
             "opera_android": {
               "version_added": "15",
-              "version_removed": "43"
+              "version_removed": "15"
```
Comment (Elchi3) [inline api/AudioListener.json]: 43 looks correct to me.

File: api/AudioListener.json [hunk 4]
```diff
@@ -461,7 +461,7 @@
...
```
Comment (Elchi3) [inline api/AudioListener.json]: 43 seems correct to me.

File: api/AudioParam.json [hunk 5]
```diff
@@ -305,11 +305,11 @@
             ],
             "opera_android": [
               {
-                "version_added": "44"
+                "version_added": "43"
               },
               {
                 "version_added": true,
-                "version_removed": "43",
+                "version_removed": "10.1",
```
Comment (Elchi3) [inline api/AudioParam.json]: 43 is correct I think

File: api/AudioScheduledSourceNode.json [hunk 6]
```diff
@@ -65,11 +65,11 @@
           ],
           "opera_android": [
             {
-              "version_added": "44"
+              "version_added": "43"
             },
             {
               "version_added": "15",
-              "version_removed": "43",
+              "version_removed": "15",
```
Comment (Elchi3) [inline api/AudioScheduledSourceNode.json]: 43 is correct I believe

File: api/AudioBufferSourceNode.json (hunk 1 above)
Comment (ExE-Boss) [reply api/AudioBufferSourceNode.json]: I had a bug in the script 🤦🏻‍♂️

**Comment by ExE-Boss** [comment]
review?(@Elchi3): I’ve now addressed your review.

Let’s get this merged before more stuff breaks.

**Comment by foolip** [comment]
There are a lot of conflicts already and this will also conflict with https://github.com/mdn/browser-compat-data/pull/3658.

If there's some vetting of the API/CSS/etc. data needed, can the browsers/opera.json and browsers/opera_android.json changes be merged first? Per https://github.com/mdn/browser-compat-data/issues/591 opera_android is the only browser missing in browsers/, and I noticed this today when trying to update data for it in https://github.com/mdn/browser-compat-data/pull/3731.

**Comment by ExE-Boss** [comment]
You need to run a script that updates the data according to https://github.com/mdn/browser-compat-data/pull/1712#issuecomment-416886047.

**Comment by foolip** [comment]
@ExE-Boss I'm not setting any version numbers though, just changing null to true.

**Comment by ExE-Boss** [comment]
My script also affects the case when `opera_android` is `null` and `opera` is non&#x2011;null

**Comment by foolip** [comment]
So, do I need to do anything if I'm just changing null to true based on testing in the real browser, or is it just that there's still more work that could be done?

**Comment by ExE-Boss** [comment]
It means that there will be merge conflicts.

**Comment by vinyldarkscratch** [comment]
We'll definitely be facing merge conflicts, so the sooner we can get this merged, the better!

**Comment by Elchi3** [comment]
So, I've looked into this again and the mapping seems to be applied correctly now. Thanks for your work!

The only thing that this doesn't seem to take into account is that when mobile chromiums differ from desktop chromiums. This is quite rare, but for example, look at https://developer.mozilla.org/en-US/docs/Web/CSS/@viewport/zoom. Here, Chrome for Android is much later than Chrome desktop (I don't know if the data is correct btw) but your mapping here takes the Chrome desktop number. I would assume that Opera Android is closer to Chrome for Android and thus the Opera Android number could have been based on the Chrome Android version. 

I believe that majority of the changes in this PR is correct and I think that the above scenario is relatively minor. So, I think it would make sense to go ahead and merge this if I don't hear other thoughts on this matter.

It needs another rebase though. Sorry.

**Comment by jpmedley** [comment]
Rule of thumb: if device hardware is involved, don't be surprised if desktop and mobile have different version numbers. In fact you can almost expect it. (The hardware interaction here comes from the fact that zooming on mobile has to be done by pinching the screen.)

The difference is usually only a few versions. There seems to be a difference of 32 versions for this feature. If the difference is excessively large, ping me and I'll look into it.

**Comment by Elchi3** [comment]
Thanks for the rebase!

I've double checked again https://github.com/mdn/browser-compat-data/blob/aa314c9bce0d3dea20a08011a0f6feeb833fc52a/browsers/opera_android.json against https://github.com/mdn/browser-compat-data/pull/1712#issuecomment-416886047

In the comment I said that there weren't these releases:
17
23
31
38 - 40

But you've added them to opera_android.json. Why? Did you find evidence that these releases existed?

**Comment by ExE-Boss** [comment]
I think I forgot to remove them.

**Comment by ExE-Boss** [comment]
review?(@Elchi3): I’ve removed the non‑existent **Opera Android** versions, now would probably be a good time to merge this before more merge conflicts pile up.
